File: legacy_restore/src_local/utils/submodule_sync_logic.py
"""
SISTEMA DE PERSONAS AGENTES - SUPORTE TÉCNICO
Módulo: Lógica de Sincronia de Submódulos (SubmoduleSyncLogic)
Função: Centralizar validações de trava (lock) e deltas de commit.
"""
import time
from datetime import datetime, timedelta
import logging

logger = logging.getLogger(__name__)

class SubmoduleSyncLogic:
    def is_locked(self, lock_file):
        if not lock_file.exists(): return False
        
        mtime = datetime.fromtimestamp(lock_file.stat().st_mtime)
        if (datetime.now() - mtime) > timedelta(minutes=10):
            if lock_file.exists(): lock_file.unlink()
            return False
        return True

    def get_submodule_delta(self, git, remote):
        """Calcula a diferença de commits entre local e remoto."""
        import time
        start_time = time.time()
        
        if not remote: return []
        
        git.run(["fetch", remote], check=False)
        active = git.get_current_branch()
        tracking = git.get_tracking_branch(active)
        
        delta = git.get_commit_count(f"{active}..{remote}/{tracking}")
        
        if delta > 0:
            from src_local.utils.logging_config import log_performance
            log_performance(logger, start_time, "Telemetry: Submodule delta check")
            return [{"file": ".agent/skills", "issue": f"Delta: {delta} commits", "severity": "CRITICAL", "context": "DependencyAuditor"}]
        return []

File: src_local/utils/submodule_sync_logic.ts
import * as fs from "node:fs/promises";
import winston from "winston";
import { GitClient } from "./git_client.ts";

const logger = winston.child({ module: "SubmoduleSyncLogic" });

/**
 * 🔄 Lógica de Sincronia de Submódulos (Bun Version).
 * Centraliza validações de trava (lock) e deltas de commit.
 */
export class SubmoduleSyncLogic {
    async isLocked(lockPath: string): Promise<boolean> {
        try {
            const stats = await fs.stat(lockPath).catch(() => null);
            if (!stats) return false;

            const mtime = stats.mtime;
            const diffMinutes = (Date.now() - mtime.getTime()) / (1000 * 60);

            if (diffMinutes > 10) {
                await fs.unlink(lockPath).catch(() => { });
                return false;
            }
            return true;
        } catch {
            return false;
        }
    }

    async getSubmoduleDelta(git: GitClient, remote: string | null): Promise<any[]> {
        const startTime = Date.now();

        if (!remote) return [];

        try {
            await git.run(["fetch", remote]);
            const active = await git.getCurrentBranch();
            const tracking = await git.getTrackingBranch(active);

            const delta = await git.getCommitCount(`${active}..${remote}/${tracking}`);

            if (delta > 0) {
                const duration = (Date.now() - startTime) / 1000;
                logger.info(`Telemetry: Submodule delta check completed in ${duration.toFixed(4)}s`);
                return [{
                    file: ".agent/skills",
                    issue: `Delta: ${delta} commits pendentes no submódulo.`,
                    severity: "CRITICAL",
                    context: "DependencyAuditor"
                }];
            }
        } catch (e) {
            logger.warn(`⚠️ Erro ao verificar delta de submódulo: ${e}`);
        }

        return [];
    }
}
